### DISCUSSION TITLE
Scaffolding and Running Extensions

### DISCUSSION THREAD

**Discussion by beschneider** [directus/directus#10405] [Archive]
I am trying to get going on custom endpoints. The docs describe a utility to scaffold an extension: 

`npm init directus-extension`

**Question 1: do I run this command in the desired extension directory, i.e. extensions/endpoints?**

Assuming yes and I build an endpoint called "test," I then run `npm run build` as the docs suggest. But then when I start up the app with `npx directus start`, I get the following error:

> Couldn't register endpoint "test"
> Cannot find module '......./extensions/endpoints/test/index.js'

It seems as if Directus is looking for index.js in _test/_ instead of _test/src/_. But having the index.js file in a _src/_ directory is how the utility scaffolds the extension and how the docs suggest it should be structured.

**Question 2: am I misunderstanding the structure of an extension?**

If I move index.js up one directory out of _src/_, I then get the following error:

> Couldn't register endpoint "test"
> Unexpected token 'export'

**Comment by beschneider** [comment]
I figured this out.

Use the utility to scaffold your extension outside of the Directus app as a separate project. Then build the extension. Then copy the built extension from its _/dist_ directory over to the Directus app’s _/extensions/endpoints/name-of-endpoint/_ directory.

It might help if the documentation provided a little more context on how extension work.

**Comment by DanielBiegler** [comment]
Hi there, I also just struggled with this and solved it by simply exporting it like this:

```javascript
// /extensions/endpoints/custom/index.js

module.exports = (router, { services, exceptions, logger }) => {
    const { ItemsService } = services;
    const { ServiceUnavailableException } = exceptions;
    
    router.get('/', (req, res, next) => {
        res.json({ example: "Hello World!" })
    });

    router.get('/example', (req, res, next) => {
        const exampleService = new ItemsService("your_collection", { schema: req.schema });

        exampleService.readByQuery({ fields: ["*"] })
        .then(results => res.json(results))
        .catch(error => next(new ServiceUnavailableException(error.message)));
    });

}
```

I first tried the way @beschneider described via the directus scaffolding package - which works. I looked at the file and noticed that the output is just simply built into a function that gets exported via `module.exports`.

I guess if you're interested in sharing your extension with others the scaffolding approach can be nice. For me personally having one `index.js` is enough for now. (This also works with hooks)

**Reply by azrikahar** [reply]
@DanielBiegler we'll indeed need to use commonjs syntax with `module.exports` for endpoints, as well as hooks 👍 Just to share more info and clarify any confusion on why sometimes it's ES6 (`export default`) or CommonJS (`module.exports`) syntax, you can check out this comment: https://github.com/directus/directus/discussions/10656#discussioncomment-1863343. Hope it helps!
